keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,960 @@
+running containers/pods as non-root.
+
+Running containers as non-root has been a long recommended best-practice in 
+kubernetes and we have published [blogs](https://kubernetes.io/blog/2018/07/18/11-ways-not-to-get-hacked/#8-run-containers-as-a-non-root-user) recommending this best practice. kubeadm
+which is a tool built to provide a "best-practice" path to creating clusters is
+an ideal candidate to apply this to.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Run control-plane components as non-root in kubeadm. More specifically :-
+  - Run `kube-apiserver` as non-root in kubeadm.
+  - Run `kube-controller-manager` as non-root in kubeadm.
+  - Run `kube-scheduler` as non-root in kubeadm.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Run node components as non-root in kubeadm. More specifically :-
+  - Run `kube-proxy` as non-root.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Here is what we propose to do at a high level:
+
+1. Run the control-plane components as non-root, by assigning a unique uid/gid to the containers in the Pods using the `runAsUser` and `runAsGroup` fields in the [securityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) on the Pods.
+
+2. Drop all capabilities from kube-controller-manager and kube-scheduler Pods. Drop all but cap_net_bind_service from kube-apiserver Pod, this will allow us to run as non-root while still being able to bind to ports < 1024. This can be achieved by using the `capabilities` field in the [securityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) of the containers in the Pods.
+
+3. Set the seccomp profile to `runtime/default`, this can be done by setting the `seccompProfile` field in the `securityContext`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
i'd say having the cluster and individual component credentials on disk as non-root owned holds a number of risks.
how do we prevent attacks?

this should be covered here...

Reply:
What risk does having the credentials as non-root hold? The only risk I can think of is that if something is running as root it can read those credentials. Running each component with different ids and setting the file permissions for the non-shared files to only that id, better isolates the files.

Reply:
> What risk does having the credentials as non-root hold? 

that is a question the KEP should pose and answer.
we don't want the contents of /etc/kubernetes/ to be readable by unwanted users/groups.

Reply:
e.g. We set a UID and GID of an actually existing user/group. This is most likely in environments where the machine is joined to some external directory (OpenLDAP or Active Directory), and UIDs/RIDs are hashes of some other attribute (as winbind does in SAMBA). If we follow using `adduser --system` or using the appropriate ranges from /etc/login.defs, that should act as our mitigation.

Reply:
Done.